detection_rules/misc.py

```diff
@@ -32,6 +55,291 @@
+                    if release_name and release_name == r.title or tag_name and tag_name == r.tag_name), None)
+
+    return release
+
+
+def upload_gh_release_asset(token, asset: bytes, asset_name, repo=None, content_type='application/zip',
+                            release_name=None, tag_name=None, upload_url=None):
+    """Save a Github relase asset."""
+    if not upload_url:
+        assert repo, 'You must provide a repo name if not providing an upload_url'
+
+        release = get_gh_release(repo, release_name, tag_name)
+        upload_url = release['upload_url']
+        suffix = '{?name,label}'
+        upload_url = upload_url.replace(suffix, f'?name={asset_name}&label={asset_name}')
+
+    headers = {'content-type': content_type}
+    r = requests.post(upload_url, auth=('', token), data=asset, headers=headers)
+    r.raise_for_status()
+    click.echo(f'Uploaded {asset_name} to release: {r.json()["url"]}')
+
+
+def load_zipped_gh_assets_with_metadata(url) -> Tuple[str, dict]:
+    """Download and unzip a GitHub assets."""
+    response = requests.get(url)
+    zipped_asset = ZipFile(io.BytesIO(response.content))
+    zipped_sha256 = hashlib.sha256(response.content).hexdigest()
+
+    assets = {}
+    for zipped in zipped_asset.filelist:
+        if zipped.is_dir():
+            continue
+
+        contents = zipped_asset.read(zipped.filename)
+        sha256 = hashlib.sha256(contents).hexdigest()
+
+        assets[zipped.filename] = {
+            'contents': contents,
+            'metadata': {
+                'compress_size': zipped.compress_size,
+                # zipfile provides only a 6 tuple datetime; -1 means DST is unknown;  0's set tm_wday and tm_yday
+                'created_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', zipped.date_time + (0, 0, -1)),
+                'sha256': sha256,
+                'size': zipped.file_size,
+            }
+        }
+
+    return zipped_sha256, assets
+
+
+def load_json_gh_asset(url) -> dict:
+    """Load and return the contents of a json asset file."""
+    response = requests.get(url)
+    response.raise_for_status()
+    return response.json()
+
+
+def download_gh_asset(url, path, overwrite=False):
+    """Download and unzip a GitHub asset."""
+    zipped = requests.get(url)
+    z = ZipFile(io.BytesIO(zipped.content))
+
+    Path(path).mkdir(exist_ok=True)
+    if overwrite:
+        shutil.rmtree(path, ignore_errors=True)
+
+    z.extractall(path)
+    click.echo(f'files saved to {path}')
+
+    z.close()
+
+
+class GithubClient:
+    """GitHub client wrapper."""
+
+    def __init__(self, token=None):
+        """Get an unauthenticated client, verified authenticated client, or a default client."""
+        if not Github:
+            raise ModuleNotFoundError('Missing PyGithub - try running `pip install -r requirements-dev.txt`')
+
+        self.client: Github = Github(token)
+        self.unauthenticated_client = Github()
+        self.__token = token
```

Comment:
if `self.client` is public and has `token`, does making this private solve anything extra?

Reply:
makes it a few layers deeper at least. Plus there is no need for direct access to it, so it isn't hurting anything either